lib/hanami/container.rb

```diff
@@ -0,0 +1,115 @@
+# frozen_string_literal: true
+
+require "dry/system/container"
+
+module Hanami
+  # Hanami private IoC
+  #
+  # @since 2.0.0
+  class Container < Dry::System::Container
+    boot(:root) do
+      start do
+        register(:root, Hanami.root)
+      end
+    end
+
+    boot(:env) do |c|
+      init do
+        use :root
+
+        begin
+          require "dotenv"
+        rescue LoadError # rubocop:disable Lint/HandleExceptions
+        end
+
+        Dotenv.load(c[:root].join(".env")) if defined?(Dotenv)
+      end
+    end
+
+    boot(:lib) do |c|
+      init do
+        use :root
+
+        $LOAD_PATH.unshift root.join("lib")
+        Hanami::Utils.require!(c[:root].join("lib", "**", "*.rb"))
+      end
+    end
+
+    boot(:configuration) do |c|
+      init do
+        use :root
+
+        require c[:root].join("config", "application").to_s
+      end
+
+      start do
+        register(:configuration, Hanami.application.configuration.finalize)
+      end
+    end
```

Comment:
Why `configuration` is stored in the class instead of the instance?

Reply:
It's stored at the class level to allow the _config DSL_. Example:

```ruby
module Soundeck
  class Application < Hanami::Application
    # This corresponds to Hanami.application.configuration
    config.whatever = :value
  end
end
```

As you know, I'm trying my best to remove class configurations and global state, but some glue code has still exist in a few spots to make it work.

Reply:
> It's stored at the class level to allow the config DSL

Ah ok, so is it `config` vs `configuration`? Maybe this should be always `config`.

> As you know, I'm trying my best to remove class configurations and global state, but some glue code has still exist in a few spots to make it work.

That's understandable!

Reply:
It's mainly `configuration`, but it's aliased to save some keystrokes to devs. https://github.com/hanami/hanami/pull/975/files#diff-00f27089d3c7dbf5783b4a35405b603aR36